opennav_following/src/following_server.cpp

```diff
@@ -0,0 +1,676 @@
+      checkAndWarnIfPreempted<FollowObject>(following_action_server_, "follow_object"))
+    {
+      return false;
+    }
+
+    // If we have a target frame_id from the goal, use that instead of pose detection
+    if (!target_frame.empty()) {
+      if (!getFramePose(target_frame, object_pose)) {
+        throw opennav_docking_core::FailedToDetectDock(
+          "Failed to get pose in target frame: " + target_frame);
+      }
+    } else {
+      // Otherwise, use the traditional pose detection from topic
+      if (!getRefinedPose(object_pose)) {
+        throw opennav_docking_core::FailedToDetectDock("Failed object detection");
+      }
+    }
+
+    // Get the pose at the distance we want to maintain from the object
+    // Stop and report success if goal is reached
+    auto target_pose = getPoseAtDistance(object_pose, desired_distance_);
+    if (isGoalReached(target_pose)) {
+      return true;
+    }
+
+    // The control law can get jittery when close to the end when atan2's can explode.
+    // Thus, we reduce the desired distance by a small amount so that the robot never
+    // gets to the end of the spiral before its at the desired distance to stop the
+    // following procedure.
+    const double backward_projection = 0.25;
+    const double effective_distance = desired_distance_ - backward_projection;
+    target_pose = getPoseAtDistance(object_pose, effective_distance);
+
+    // ... and transform the target_pose into base_frame
+    try {
+      tf2_buffer_->transform(
+        target_pose, target_pose, base_frame_, tf2::durationFromSec(transform_tolerance_));
+    } catch (const tf2::TransformException & ex) {
+      RCLCPP_WARN(get_logger(), "Failed to transform target pose: %s", ex.what());
+      return false;
+    }
+
+    // If the object is behind the robot, we reverse the control
+    auto robot_pose = getRobotPoseInFrame(target_pose.header.frame_id);
+    const double dist = std::hypot(
+      robot_pose.pose.position.x - target_pose.pose.position.x,
+      robot_pose.pose.position.y - target_pose.pose.position.y);
+    bool reversing = allow_backward_ &&
+      dist < desired_distance_ && target_pose.pose.position.x < robot_pose.pose.position.x;
+
+    // Compute and publish controls
+    auto command = std::make_unique<geometry_msgs::msg::TwistStamped>();
+    command->header.stamp = now();
+    if (!controller_->computeVelocityCommand(target_pose.pose, command->twist, true, reversing)) {
+      throw opennav_docking_core::FailedToControl("Failed to get control");
+    }
+    vel_publisher_->publish(std::move(command));
+
+    loop_rate.sleep();
+  }
+  return false;
+}
+
+bool FollowingServer::rotateToObject(geometry_msgs::msg::PoseStamped & object_pose)
+{
+  const double dt = 1.0 / controller_frequency_;
+  auto target_pose = object_pose;
+  target_pose.pose.orientation = nav2_util::geometry_utils::orientationAroundZAxis(
+    tf2::getYaw(target_pose.pose.orientation) + M_PI);
```

Comment:
So PI makes it turn 180, but always turns one way. As such, we don't fully explore the space. Should we do:
* A full rotation?
* Do a 90 degree left and then 180 deg right (so we do search 90 degrees left and right)? Kind of a wiggle move